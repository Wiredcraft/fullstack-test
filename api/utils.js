const sumVotes = votes => {
  if (votes) {
    return votes.reduce((acc, cur) => {
      return acc + cur.val;
    }, 0);
  } else {
    return 0;
  }
};

module.exports = {
  sumVotes
};
