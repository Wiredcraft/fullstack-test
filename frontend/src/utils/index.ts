import * as dayjs from 'dayjs'
import relativeTime from 'dayjs/plugin/relativeTime';
dayjs.extend(relativeTime);

export function formatTime(date: Date) {
  return dayjs(date).fromNow()
}