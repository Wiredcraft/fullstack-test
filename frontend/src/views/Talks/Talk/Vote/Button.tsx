/* eslint-disable arrow-body-style */
/* eslint-disable @typescript-eslint/explicit-function-return-type */

import React, {ReactElement} from 'react';

interface VoteButtonProps {
    id: string;
    children: React.ReactNode;
    onClick: Function;
    vote: number;
  }

const VoteButton = (props: VoteButtonProps): ReactElement => {
  return (
    <div style={{
      cursor: 'pointer',
    }} onClick={() => props.onClick(props.id, props.vote)}>
      {props.children}
    </div>
  );
};

export default VoteButton;
