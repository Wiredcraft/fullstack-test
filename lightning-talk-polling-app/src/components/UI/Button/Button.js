import React from 'react';

const button = (props) => (
    <button
        className={props.cssClass}
        disabled={props.disabled || false}
        onClick={props.clicked}
    >
        {props.text}
    </button>
);

export default button;