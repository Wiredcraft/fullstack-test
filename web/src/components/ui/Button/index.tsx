import React, { ButtonHTMLAttributes, FC } from "react";
import styled, { css } from "styled-components";
import { rgba } from "polished";
import * as tokens from "../../../styles/tokens";

type ButtonType = "primary" | "translucent";
type ButtonSize = "small" | "middle" | "large";

const translucent = css`
  border: none;
  background: ${({ theme: { colors } }) => rgba(colors.primary.base, 0.12)};
  color: ${({ theme: { colors } }) => colors.primary.base};
`;
const primary = css`
  color: ${tokens.colors.white};
  border: none;
  background: ${({ theme: { colors } }) => colors.primary.base};
`;

const buttonCSS = {
  translucent,
  primary
};
const sizes = {
  small: 24,
  middle: 32,
  large: 40
};

interface Props {
  variant?: ButtonType;
  size?: ButtonSize;
}

const Container = styled.button<Required<Props>>`
  border-radius: ${({ size }) => `${sizes[size] / 2}px`};
  font-weight: 500;
  font-variant: all-small-caps;
  font-size: ${({ size }) => `${sizes[size] / 2}px`};
  height: ${({ size }) => `${sizes[size]}px`};
  min-width: ${({ size }) => `${sizes[size] * 2}px`}
  padding: ${({ size }) => `0 ${sizes[size] / 2}px`};
  cursor: pointer;
  outline: none;
  ${({ variant }) => buttonCSS[variant]};
  &:disabled {
   opacity: 0.6;
   cursor: not-allowed;
  }
`;

const Button: FC<Props & ButtonHTMLAttributes<HTMLButtonElement>> = ({
  variant = "primary",
  size = "middle",
  children,
  ...otherProps
}) => {
  return (
    <Container variant={variant} size={size} {...otherProps}>
      {children}
    </Container>
  );
};

export default Button;
