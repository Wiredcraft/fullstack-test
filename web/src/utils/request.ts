import axios from "axios";
import { STORE_USER_INFO } from "../constants/auth";

const BASE_URL = "http://localhost:8383";
const instance = axios.create({
  baseURL: BASE_URL,
  timeout: 5000
});

instance.interceptors.request.use(config => {
  const { headers = {} } = config;
  const userString = localStorage.getItem(STORE_USER_INFO);
  if (userString) {
    try {
      const { token } = JSON.parse(userString);
      if (token) {
        headers.Authorization = `Bearer ${token}`;
      }
      // eslint-disable-next-line no-empty
    } catch (e) {}
  }
  return config;
});

instance.interceptors.response.use(
  response => {
    if (response.data.success) {
      return response.data.result;
    }
    throw response.data.reason;
  },
  error => Promise.reject(error)
);

export default instance;
