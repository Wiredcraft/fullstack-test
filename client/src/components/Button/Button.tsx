import { ButtonHTMLAttributes } from 'react';

import style from './style.css';

import { createStyler } from '@/utils/styler';

const styler = createStyler(style);

interface Props extends ButtonHTMLAttributes<Element> {
  className?: string;
  children?: React.ReactNode;
  shape?: 'filled' | 'outline' | 'link';
}

export function Button(props: Props) {
  const { className, children, shape = 'link', ...buttonProps } = props;

  return (
    <button className={styler(className, 'button', `${shape}-button`)} {...buttonProps}>
      {children}
    </button>
  );
}
