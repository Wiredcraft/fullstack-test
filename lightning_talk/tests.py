from django.contrib.auth.models import User
from rest_framework import status
from rest_framework.test import APITestCase
from rest_framework_jwt.utils import jwt_payload_handler, jwt_encode_handler
from .models import Post, Upvote


class PostTest(APITestCase):
    def setUp(self):
        self.user = User.objects.create_user(
            username='casca', password='pwd@1234')
        payload = jwt_payload_handler(self.user)
        self.token = jwt_encode_handler(payload)

    def test_create_post(self):
        """
        Create a new post object.
        """
        url = '/lightning_talk/api/posts/'
        data = {'title': 'yo', 'description': 'hei'}
        response = self.client.post(
            url, data, format='json', HTTP_AUTHORIZATION='JWT ' + self.token)
        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
        self.assertEqual(Post.objects.count(), 1)
        self.assertEqual(Post.objects.get().description, 'hei')

    def test_update_post(self):
        """
        Update a post object
        """
        post = Post.objects.create(
            title='yo', description='hei', user=self.user)

        url = '/lightning_talk/api/posts/%s/' % post.pk
        data = {'description': 'hei ho'}
        response = self.client.patch(
            url, data, format='json', HTTP_AUTHORIZATION='JWT ' + self.token)
        self.assertEqual(response.status_code, status.HTTP_200_OK)
        self.assertEqual(Post.objects.get().description, 'hei ho')

    def test_delete_post(self):
        """
        Delete a post object
        """
        post = Post.objects.create(
            title='yo', description='hei', user=self.user)

        url = '/lightning_talk/api/posts/%s/' % post.pk
        response = self.client.delete(
            url, HTTP_AUTHORIZATION='JWT ' + self.token)
        self.assertEqual(response.status_code, status.HTTP_204_NO_CONTENT)
        self.assertEqual(Post.objects.count(), 0)

    def test_upvote_post(self):
        """
        Upvote post
        """
        post = Post.objects.create(
            title='yo', description='hei', user=self.user)

        url = '/lightning_talk/api/posts/%s/upvote/' % post.pk
        response = self.client.put(url, HTTP_AUTHORIZATION='JWT ' + self.token)
        self.assertEqual(response.status_code, status.HTTP_200_OK)
        self.assertEqual(Post.objects.get().upvotes.count(), 1)

    def test_cancle_upvote_of_post(self):
        """
        Cancle upvote post
        """
        post = Post.objects.create(
            title='yo', description='hei', user=self.user)
        upvote = Upvote.objects.create(user=self.user, post=post)
        post.upvotes.add(upvote)

        url = '/lightning_talk/api/posts/%s/upvote/' % post.pk
        response = self.client.delete(
            url, HTTP_AUTHORIZATION='JWT ' + self.token)
        self.assertEqual(response.status_code, status.HTTP_200_OK)
        self.assertEqual(Post.objects.get().upvotes.count(), 0)
