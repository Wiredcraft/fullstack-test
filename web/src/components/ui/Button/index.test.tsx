import React from "react";
import "jest-styled-components";
import { render } from "../../../../tests/utils";
import "../../../../tests/mock";
import Button from ".";

describe("test <Button />", () => {
  it("should match snapshot", () => {
    const { container } = render(<Button variant="primary" />);
    expect(container.firstChild).toMatchSnapshot();
  });
  it("should match snapshot", () => {
    const { container } = render(<Button variant="translucent" />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
