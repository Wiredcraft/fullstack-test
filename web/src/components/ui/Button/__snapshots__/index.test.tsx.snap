// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <Button /> should match snapshot 1`] = `
.c0 {
  border-radius: 16px;
  font-weight: 500;
  font-variant: all-small-caps;
  font-size: 16px;
  height: 32px;
  min-width: 64px;
  padding: 0 16px;
  cursor: pointer;
  outline: none;
  color: #FFFFFF;
  border: none;
  background: #3B66F5;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

<button
  class="c0"
/>
`;

exports[`test <Button /> should match snapshot 2`] = `
.c0 {
  border-radius: 16px;
  font-weight: 500;
  font-variant: all-small-caps;
  font-size: 16px;
  height: 32px;
  min-width: 64px;
  padding: 0 16px;
  cursor: pointer;
  outline: none;
  border: none;
  background: rgba(59,102,245,0.12);
  color: #3B66F5;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

<button
  class="c0"
/>
`;
